### PR TITLE
#201 (feat): added envId to the storage key and added property to cacheOptions to customize the key

### PR DESCRIPTION
**Pull request by oluizcarvalho**
### Problem:
A limitation was found in the `clientSide cacheFlags` implementation. The issue arises because the same key is always used in `localStorage` cited in #201, which can lead to conflicts when the system is used in the context of micro frontends (MFEs) or with multiple instances. This key competition in `localStorage` can cause negative effects, such as data being overwritten or cache inconsistencies.

### Resolution:
To address this issue, it was proposed to allow the `localStorage` key name to be explicitly defined or for it to take into account the `environmentID`. This would solve the problem by ensuring that each instance or environment has a unique key, preventing conflicts.

### Changes Implemented:
1. **Concatenating `environmentID` to the `localStorage` Key:** To resolve the initial problem, the key used in `localStorage` was modified to include the `environmentID`. This ensures that different environments use distinct keys, eliminating competition in local storage.

2. **New `storageKey` Property in `cacheOptions` Object:** During the process, an additional issue was identified: two or more instances with the same `environmentID` but identifying different users could still cause conflicts. To handle this scenario and increase flexibility, the `storageKey` property was added to the `cacheOptions` object, which is passed during initialization. This allows the `localStorage` key to be customized for each instance, ensuring the necessary distinction between different usage contexts.

### Important Note:
When deploying the new version of the SDK with these changes, there may be a loss of compatibility with the old key used to retrieve the cache. It’s important to consider this when upgrading, as the old cache data may not be accessible using the new key format.

### Conclusion:
These changes enhance the flexibility and robustness of the system when handling cache in complex environments, such as applications using micro frontends or multiple instances. Now, conflicts in `localStorage` can be avoided, and the cache can function correctly for each specific context.

> All tests were run successfully

### CODE REVIEW

**Comment by matthewelwell** [comment]
Thanks for this @oluizcarvalho , we'll review it as soon as we can!

**Comment by kyle-ssg** [comment]
I appreciate all the considerations here, including that it's a breaking change for the next cache hit, there's definitely no easy way to solve this.

I just wonder, if we really wanted to make this backwards compatible we could check both locations I guess, prioritising the environment version. 

 Also, I wonder what the behaviour of this would be if the environment id was switched, I've heard of people doing this and it could be related to #236

**Comment by kyle-ssg** [comment]
Ok after considering it internally, lets go ahead with this. Minor comment as a result 😄

File: flagsmith-core.ts [hunk 1]
```diff
@@ -37,8 +37,10 @@ type RequestOptions = {
 }
 
 let AsyncStorage: AsyncStorageType = null;
-const FLAGSMITH_KEY = "BULLET_TRAIN_DB";
-const FLAGSMITH_EVENT = "BULLET_TRAIN_EVENT";
+const DEFAULT_FLAGSMITH_KEY = "BULLET_TRAIN_DB";
```
Comment (kyle-ssg) [inline flagsmith-core.ts:40]: Let's (finally) change this to FLAGSMITH!

**Comment by oluizcarvalho** [comment]
> I appreciate all the considerations here, including that it's a breaking change for the next cache hit, there's definitely no easy way to solve this.
> 
> I just wonder, if we really wanted to make this backwards compatible we could check both locations I guess, prioritising the environment version.
> 
> Also, I wonder what the behaviour of this would be if the environment id was switched, I've heard of people doing this and it could be related to #236

I had thought of a solution like this for the problem of backwards compatibility of keys
```ts
const migrateAsyncStorageCache = async () => {
                try {
                    // Check if data exists under the old key using getItemSync or getItem
                    const oldData = AsyncStorage?.getItemSync ?
                        AsyncStorage.getItemSync(oldKey) :
                        await AsyncStorage?.getItem(oldKey);

                    if (oldData) {
                        // Migrate the data to the new key
                        await AsyncStorage?.setItem(newKey, oldData);

                        // Remove the old key after migration
                        await AsyncStorage?.removeItem(oldKey);

                        console.log(`Cache migrated from '${oldKey}' to '${newKey}'.`);
                    }
                } catch (error) {
                    console.error(`Failed to migrate cache from '${oldKey}' to '${newKey}':`, error);
                }
            }
```
It would be a solution that we can solve in the future and would help us with migration without breaking changes, what do you think?

File: flagsmith-core.ts (hunk 1 above)
Comment (oluizcarvalho) [reply flagsmith-core.ts]: resolved

**Comment by kyle-ssg** [comment]
> > Also, I wonder what the behaviour of this would be if the environment id was switched, I've heard of people doing this and it could be related to #236
> 
> Excellent point, in fact we have to worry about environmentID changes, today from what I've seen we only support one envId at a time in storage and I honestly think it makes sense, what we could implement would be a format to verify that when there are changes we clean the old storage referring to the old envId

No I think there are cases where we want multiple, e.g. there are some fragmented frontends that fetch flags from completely different projects.

**Comment by kyle-ssg** [comment]
> > I appreciate all the considerations here, including that it's a breaking change for the next cache hit, there's definitely no easy way to solve this.
> > I just wonder, if we really wanted to make this backwards compatible we could check both locations I guess, prioritising the environment version.

To be honest, I think lets keep it without, it introduces tech debt and really this new version would just introduce a single cache miss.

**Comment by oluizcarvalho** [comment]
> > It would be a solution that we can solve in the future and would help us with migration without breaking changes, what do you think?
> > 
> > Another solution would be to not change the default key and leave only the second part implemented, which is the optional new parameter.
> 
> To be honest, I think lets keep it without this new change, it introduces tech debt and really this new version would just introduce a single cache miss.
> 
> 

I agree!

**Comment by oluizcarvalho** [comment]
@kyle-ssg What is the next step to move forward with PR?

**Comment by oluizcarvalho** [comment]
@kyle-ssg What is the next step to move forward with PR?

**Comment by kyle-ssg** [comment]
@oluizcarvalho I'll do a bit of testing with this but I don't see why we wouldn't be releasing a new major version most likely tomorrow.

File: flagsmith-core.ts [hunk 2]
```diff
@@ -393,6 +399,27 @@ const Flagsmith = class {
                 }
             }
 
+            const migrateAsyncStorageCache = async () => {
```
Comment (kyle-ssg) [inline flagsmith-core.ts:402]: We can remove this as a result of our conversation right ?

Comment (oluizcarvalho) [reply flagsmith-core.ts:402]: Of course, I hadn't noticed that this code was in my last commit, thanks

**Comment by oluizcarvalho** [comment]
@kyle-ssg resolvidos os conflitos do merge

**Comment by oluizcarvalho** [comment]
Hey @kyle-ssg, we can prepare the deploy or maybe a pre-release for me to test in my projects?

**Comment by kyle-ssg** [comment]
Hey @oluizcarvalho yeah let's do that! I'll get this merged into a feature branch on flagsmith / bump the versions

**Pull request merged**